### PR TITLE
add delete_oldest_seg func for topic module

### PR DESCRIPTION
**Pull request by redink**

### CODE REVIEW

File: src/gululog_idx.erl [hunk 1]
```diff
@@ -121,12 +122,24 @@ append(#idx{ version = ?LOGVSN
 %%    should be done using truncate API instead.
 %% @end
 -spec delete_oldest_seg(dirname(), index()) -> {segid() | false, index()}.
-delete_oldest_seg(Dir, #idx{tid = Tid, segid = CurrentSegId} = Index) ->
+delete_oldest_seg(Dir, Index) ->
+  delete_oldest_seg(Dir, Index, ?undef).
```
Comment (redink) [inline src/gululog_idx.erl]: Add argument for backup, and keep the delete_oldest_seg/2 function.

File: src/gululog_idx.erl [hunk 2]
```diff
@@ -542,6 +555,21 @@ gululog_idx_test_() ->
...
         end
       }
+    , { "delete oldest seg"
+      , fun() ->
+          Dir = test_dir(),
+          BackupDir = filename:join(Dir, "backup"),
+          Idx0 = init(Dir),
+          ?assertEqual({0, Idx0}, delete_oldest_seg(Dir, Idx0)),
+          DeleteFile1 = mk_name(Dir, 0),
+          ?assertEqual(false, lists:member(DeleteFile1, wildcard_reversed(Dir))),
```
Comment (redink) [inline src/gululog_idx.erl:556]: Ensure the file is deleted.

File: src/gululog_topic.erl [hunk 3]
```diff
@@ -126,6 +128,34 @@ truncate(#topic{dir = Dir, idx = Idx, cur = Cur} = Topic, LogId, BackupDir) ->
        IdxFileOpList ++ SegFileOpList}
   end.
 
+%% @doc Delete oldest segment
+%% return the segid that is deleted, return 'false' in case:
+%% 1. nothing to delete
+%% 2. the oldest is also the latest
+%% @end
+-spec delete_oldest_seg(topic()) -> {false | segid(), topic()}.
```
Comment (zmshi) [inline src/gululog_topic.erl]: return [file_op()] instead.
REASON: this is mostly for higher level callers to log events etc. we can perhaps log it (using lager for example) here, but, It's probably a bad idea to add a logging app dependency in gululog level.

File: src/gululog_w_cur.erl [hunk 4]
```diff
@@ -106,6 +108,22 @@ truncate(Dir, Cur, SegId, Position, BackupDir) ->
...
+-spec delete_seg(dirname(), segid()) -> segid().
+delete_seg(Dir, SegId) ->
+  delete_seg(Dir, SegId, ?undef).
+
+%% @doc Delete segment file depends on given segid
+%% return the deleted segid
+%% backup the segment file when the given argument 'backupdir' is not '?undef'
+%% @end
+-spec delete_seg(dirname(), segid(), ?undef | dirname()) -> segid().
+delete_seg(Dir, SegId, BackupDir) ->
```
Comment (zmshi) [inline src/gululog_w_cur.erl]: should accept the #wcur{} structure as the second argument, two reasons:
1. We should assert the the one being deleted is not Cur#wcur.segid
2. To force the API usage for gululog_topic only.

**Comment by zmshi** [comment]
fixes issue #16

**Pull request merged**